.github/Guide.md

```diff
@@ -0,0 +1,191 @@
+* Fallback Group images are listed in [config/fallback-group.json](./config/fallback-group.json) and defined using [images from Docker Hub](https://hub.docker.com/_/node). They include OS and Node version combinations that *can* build for source.
+* Prebuilt Group images are listed in [config/prebuilt-group.json](./config/prebuilt-group.json) and defined using [images from Docker Hub](https://hub.docker.com/_/node) or using local Dockerfiles. They include OS and Node version combinations that *can not* build for source and thus require a prebuilt tarball.
+* Target Group images are listed in [config/target-group.json](./config/target-group.json) and defined using [images from Docker Hub](https://hub.docker.com/_/node) or using local Dockerfiles. They include a wide variety of OS and Node version combinations. Group includes both images that can build from code as well as those which can not.
+
+### Adding a local Dockerfile
+
+1. Find or Create a folder under `docker-node` named for the version of node the Dockerfile is created for (e.g. `16`).
+2. Create a folder to represent the operating system. Any string is valid. Use something descriptive (e.g. `ubuntu20.04.2`).
+3. Place a Docker file in the directory.
+
+### Modifying group lists
+
+1. Find available tags at: https://hub.docker.com/_/node or use local Dockerfile path (e.g. `./.github/docker-node/10/ubuntu20.04.2`)
+2. Add to appropriate group json file in `config`.
+
+### Adding a Node Version
+
+1. Create a folder under `docker-node` named for the version of node added.
+2. Create an `alpine` builder image and a `centos` builder image. Use previous node version Dockerfiles as guide.
+3. Add to appropriate group json files in `config`.
+
+### Remove a node version
+
+1. Remove version images from appropriate group json file in `config`.
+2. Leave the directory in `docker-node` for future reference.
+
+## Implementation
+
+> **tl;dr** No Actions used. Simple script gets image, returns running container id. Steps exec on container.
+
+### Workflows
+
+1. All workflows `runs-on: ubuntu-latest`.
+2. Workflows only use `actions/checkout@v2`. No other actions are used. For maintainability and security custom actions are avoided and shell scripts preferred.
+3. All jobs start with a two-step process:
+  - check out code:
+    ```yaml
+     - name: Checkout ${{ github.ref }}
+       uses: actions/checkout@v2
+    ```
+  - create a container loaded with repo code:
+    ```yaml
+      - name: Create Container  ${{ matrix.image }}
+        id: create
+        run: .github/scripts/container-from-matrix-input.sh ${{ matrix.image }}
+    ```
+4. All job steps are named.
+5. All following steps are executed using `docker exec` against the created container.
+6. All jobs end with stopping the container:
+  ```yaml
+      - name: Stop container
+        if: always()
+        run: docker stop ${{ steps.create.outputs.containerId }}
```

Comment:
This question is the crux of this refactor. 
There is no real *magic* here, but some *slight of hand* is indeed required. 

This is how it goes:
1. We need to build on very specific node/os combinations (e.g. [node:12-alpine3.9](https://github.com/nodejs/docker-node/blob/31246f5f779cafa0930a1db04bd00d875d6a940d/12/alpine3.12/Dockerfile)). We thus have to be able to define our own image, which we do via a Dockerfile (e.g. [./.github/docker-node/12/alpine3.9-build](https://github.com/appoptics/appoptics-bindings-node/blob/AO-20081-Refactor-GitHub-Actions-Test-Publish-Workflows/.github/docker-node/12/alpine3.9-build/Dockerfile)).
2. We have extensive testing needs for the package install process and build artifacts, and we prefer to use official node images for that (where available).
3. We can not currently use @actions `container` directive because it is a [wrapper over `docker create`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idcontaineroptions). We need `docker build` so we can use both pulled and built images in same matrix run.
4. We made a decision to not develop custom JS actions at this point in time. Maintainability over fun.
5. Given the above an implementation was chosen so that it can be easily described in 21 lines of documentation and enforced with *Canadian rigor* on all workflows.
6. Separation of concerns was added (i.e. moving all image group data out of YAML and into JSONs) because *sanity*.

I hope this answers all comments/questions above.